### DISCUSSION TITLE
How to run pretasks before the prompt is run

### DISCUSSION THREAD

**Discussion by jdmevo123** [microsoft/teams-sdk#1745] [Announcements]
Im trying to use the teams-ai library for a teams bot.

Im stuck trying to get a pretask run before the message is sent to the LLM by the planner.

the flow im looking for the onmessage functionality in bot-sdk.

I am trying to have:
incomming message >> Fetch data from database >> supplement incoming prompt with data >> send to LLM >> return response to user.

I have tried with planner.plan, app.run and nothing is proving to be successful in this scenario. Do you hvae any examples where a function is run to summplement the incoming request from the user before going out to the LLM?

**Comment by corinagum** [comment]
Have you looked into `beforeTurn`? 
https://github.com/microsoft/teams-ai/blob/ab2130f71fe44436ee1cd9b5904d6fa16c202dbb/getting-started/CONCEPTS/PROMPTS.md?plain=1#L104

https://github.com/microsoft/teams-ai/blob/ab2130f71fe44436ee1cd9b5904d6fa16c202dbb/getting-started/CONCEPTS/APPLICATION.md?plain=1#L89

The following is deprecated code, so it might be out of date, but it's an example of `beforeTurn` being used in a deprecated sample we had, quest bot. 
https://github.com/microsoft/teams-ai/blob/e7dcc209aa9a8633fb8a483a260227eb4ecb6b81/js/samples/04.q.questBot/src/bot.ts#L139
